elytron/WFLY-14074-normalization-of-principal-propagation.adoc

```diff
@@ -0,0 +1,107 @@
+
+If Elytron was used before the issue WFLY-12537 was resolved, then the security identity of an unsecured EJB was always anonymous if there was no incoming run-as identity.
+
+This behaviour was inconsistent with legacy behaviour. The key difference is that Elytron did not differentiate between remote and local EJB. The identity is anonymous in legacy only if mentioned unsecured EJB was called from remote context. For local EJB the identity is the current security identity obtained from security domain.
+
+This behaviour was made consistent after issue WFLY-12537 got resolved, however it would be good to provide possibility for both options and avoid regressions for users, which is what this issue is for.
+
+Before this RFE is resolved the system property could be used to switch between these behaviours.
+
+=== Summary of the two behaviours
+
+ * before WFLY-12537:
+  **  If incoming run-as identity existed then it was the identity of the EJB. Otherwise, if the EJB was secured then the identity was current security identity obtained from a security domain. If the EJB was unsecured then identity was anonymous regardless of whether the EJB was called from remote context or not.
+
+ * after WFLY-12537:
+ If incoming run-as identity exists then it is the identity of the EJB. Otherwise,
+     ** if the EJB is local then the security identity of this EJB is the current security identity obtained from security domain.
+  ** if the EJB is unsecured and called from remote context then its identity is anonymous. If the EJB is secured and called from remote context then its identity is the current security identity obtained from security domain.
+
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-14074[WFLY-14074]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1581[EAP7-1581] - EAP issue tracker
+
+=== Dev Contacts
+
+* mailto:dvilkola@redhat.com[Diana Krepinska Vilkolakova]
+
+=== QE Contacts
+
+* TODO
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE.
+// Discuss with QE during the Kickoff state to decide this
+[x] Engineering
+
+[ ] QE
+
+=== Affected Projects or Components
+
+* EJB
+* https://github.com/wildfly/wildfly[WildFly project]
+* https://github.com/wildfly-security/wildfly-elytron[WildFly Elytron project]
+
+== Requirements
+
+=== Hard Requirements
+
+* Providing possibility to configure which of the described behaviours should be used or coming to an agreement that the changes should be reverted.
+
+=== Nice-to-Have Requirements
+
+N/A
+
+=== Non-Requirements
+
+N/A
+
+== Implementation Plan
+
+The configuration option for the EJB deployments could be added to `application-security-domain` component in the `ejb3` subsystem. The parameter could be called `legacy-compliance` or more specifically `legacy-compliance-principal-injection`.
```

Comment:
Since a user would likely want to use the same behaviour (i.e., either the pre-WFLY-12537 behaviour or the post-WFLY-12537 behaviour) for all cases, instead of needing to configure this for each `application-security-domain` resource, it might be better to either introduce a new element in the EJB subsystem for this or perhaps introduce a system property to configure this.

Reply:
There is a container element `application-security-domains`. Maybe we can have an attribute under it to apply to all application-security-domain resources.

Reply:
@chengfang I will try to go that route and add it as an attribute to `application-security-domains` element. Thank you! And what do you think should be the default value for it?

Reply:
@Skyllarr I prefer the 2nd option (Have previous (legacy) behaviour by default and provide a config option that allows people to switch to the new behaviour), and clearly document it in both community doc and product doc.

Reply:
@chengfang  Thank you, yes it does make sense to use previous legacy configuration as default. I will specify this in the proposal. 

I had a conversation with @darranl about a plan to place this new configuration option `legacy-compliance-principal-propagation` as an attribute to `application-security-domains` container element. He suggested to rather place it in the child `application-security-domain` element instead, as this is kind of policy that could be applied to individual deployment. But mostly that the `application-security-domain` resources that are present in the `jb3` and `undertow` subsystem will probably be removed in the future. 

Given there are currently no configuration options in the parent `application-security-domains` resource, it might be better to not add any there before the removal. The child `application-security-domain` resource already has other configurations like `enable-jacc` and `security-domain`. So if we add `legacy-compliance-principal-propagation` option there as well, it could be handled together with those during the removal.

Reply:
@Skyllarr Can you update the implementation plan with the conclusions, and the release note content then please?

Reply:
@OndrejKotek Updated. Thanks